### PR TITLE
Feature/top p sampling

### PR DESCRIPTION
**Pull request by belerico**
This PR adds the nucleus-sampling (aka top-p sampling) as specified from https://arxiv.org/abs/1904.09751.
In top-p sampling the next token is chosen from the smallest set of tokens with a cumulative probability greater than `top-p`, i.e. by  selecting the highest probability tokens whose cumulative probability exceeds the `top-p` threshold.

### CODE REVIEW

File: litgpt/chat/base.py [hunk 1]
```diff
@@ -109,6 +116,11 @@ def main(
 
     Args:
         top_k: The number of top most probable tokens to consider in the sampling process.
+        top_p: The cumulative probability threshold to consider in the sampling process.
+            In top-p sampling the next token is sampled from the highest probability tokens
+            whose cumulative probability exceeds the threshold `top-p`.
```
Comment (rasbt) [inline litgpt/chat/base.py:121]:
```suggestion
            whose cumulative probability exceeds the threshold `top_p`.
```

File: litgpt/generate/adapter.py [hunk 2]
```diff
@@ -42,6 +43,11 @@ def main(
             for more details, see https://github.com/Lightning-AI/litgpt/blob/main/tutorials/quantize.md
         max_new_tokens: The number of generation steps to take.
         top_k: The number of top most probable tokens to consider in the sampling process.
+        top_p: The cumulative probability threshold to consider in the sampling process.
+            In top-p sampling the next token is sampled from the highest probability tokens
+            whose cumulative probability exceeds the threshold `top-p`.
```
Comment (rasbt) [inline litgpt/generate/adapter.py:48]:
```suggestion
            whose cumulative probability exceeds the threshold `top_p`.
```

File: litgpt/generate/adapter_v2.py [hunk 3]
```diff
@@ -42,6 +43,11 @@ def main(
             for more details, see https://github.com/Lightning-AI/litgpt/blob/main/tutorials/quantize.md
         max_new_tokens: The number of generation steps to take.
         top_k: The number of top most probable tokens to consider in the sampling process.
+        top_p: The cumulative probability threshold to consider in the sampling process.
+            In top-p sampling the next token is sampled from the highest probability tokens
+            whose cumulative probability exceeds the threshold `top-p`.
```
Comment (rasbt) [inline litgpt/generate/adapter_v2.py:48]:
```suggestion
            whose cumulative probability exceeds the threshold `top_p`.
```

File: litgpt/generate/base.py [hunk 4]
```diff
@@ -24,13 +24,28 @@ def multinomial_num_samples_1(probs: torch.Tensor) -> torch.Tensor:
...
+    logits: torch.Tensor, temperature: float = 1.0, top_k: Optional[int] = None, top_p: Optional[float] = None
+) -> torch.Tensor:
     logits = logits[0, -1]
     # optionally crop the logits to only the top k options
     if top_k is not None:
         v, i = torch.topk(logits, min(top_k, logits.size(-1)))
         # do not use `torch.where` as in nanogpt because it will repeat top-k collisions
         logits = torch.full_like(logits, float("-inf")).scatter_(-1, i, v)
+    # optionally crop the logits to smallest set of logits with a cumulative probability above top_p
+    if top_p is not None:
```
Comment (rasbt) [inline litgpt/generate/base.py:37]: I've always thought as `top_k` vs `top_p` not both together. Is this intentional that a user can use both, or should this raise an error if both are not `None`?

Because in theory it would be redundant if your `top_k` selects more than your `top_p` setting, right? I suppose it could be useful to have `tok_k` as a hard cut-off though if `top_p` is too generous, but in that case the if statement for `top_k` should be after the one for `top_p`?

File: litgpt/generate/base.py [hunk 5]
```diff
@@ -64,6 +80,11 @@ def generate(
         max_returned_tokens: The maximum number of tokens to return (given plus generated).
         temperature: Scales the predicted logits by 1 / temperature.
         top_k: If specified, only sample among the tokens with the k highest probabilities.
+        top_p: The cumulative probability threshold to consider in the sampling process.
+            In top-p sampling the next token is sampled from the highest probability tokens
+            whose cumulative probability exceeds the threshold `top-p`.
```
Comment (rasbt) [inline litgpt/generate/base.py:85]:
```suggestion
            whose cumulative probability exceeds the threshold `top_p`.
```

File: litgpt/generate/sequentially.py [hunk 6]
```diff
@@ -130,6 +131,11 @@ def main(
         num_samples: The number of text samples to generate.
         max_new_tokens: The number of generation steps to take.
         top_k: The number of top most probable tokens to consider in the sampling process.
+        top_p: The cumulative probability threshold to consider in the sampling process.
+            In top-p sampling the next token is sampled from the highest probability tokens
+            whose cumulative probability exceeds the threshold `top-p`.
```
Comment (rasbt) [inline litgpt/generate/sequentially.py:136]:
```suggestion
            whose cumulative probability exceeds the threshold `top_p`.
```

File: litgpt/generate/tp.py [hunk 7]
```diff
@@ -108,6 +109,11 @@ def main(
         num_samples: The number of text samples to generate.
         max_new_tokens: The number of generation steps to take.
         top_k: The number of top most probable tokens to consider in the sampling process.
+        top_p: The cumulative probability threshold to consider in the sampling process.
+            In top-p sampling the next token is sampled from the highest probability tokens
+            whose cumulative probability exceeds the threshold `top-p`.
```
Comment (rasbt) [inline litgpt/generate/tp.py:114]:
```suggestion
            whose cumulative probability exceeds the threshold `top_p`.
```

**Review by rasbt — COMMENTED** [review]
This looks great, thanks a lot! Only a few minor points from my side below:

File: litgpt/generate/base.py (hunk 4 above)
Comment (belerico) [reply litgpt/generate/base.py:37]: Yes it's intentional: i was thinking that one can decide to first select the most top-k probable tokens and then refine the selection dynamically with top-p, which is helpful since after the top-k selection the probability distribution of the next token changes with the top-p acting on the "new" distribution. I also agree that sometimes the combination of the two could return the same distribution but i think it happens for high value of top-p. 

I was also checking the[ HF default and the apply](https://github.com/huggingface/transformers/blob/9fe3f585bb4ea29f209dc705d269fbe292e1128f/src/transformers/generation/utils.py#L735):

1. Temperature scaling
2. Top-k
3. Top-p

Comment (rasbt) [reply litgpt/generate/base.py:37]: > most top-k probable tokens and then refine the selection dynamically with top-p, which is helpful since after the top-k selection the probability distribution of the next token changes with the top-p acting on the "new" distribution.

Good point, I didn't think about the rescaling after the top_k.

File: litgpt/generate/base.py [hunk 8]
```diff
@@ -24,13 +24,28 @@ def multinomial_num_samples_1(probs: torch.Tensor) -> torch.Tensor:
...
+    logits: torch.Tensor, temperature: float = 1.0, top_k: Optional[int] = None, top_p: Optional[float] = None
+) -> torch.Tensor:
     logits = logits[0, -1]
     # optionally crop the logits to only the top k options
     if top_k is not None:
         v, i = torch.topk(logits, min(top_k, logits.size(-1)))
         # do not use `torch.where` as in nanogpt because it will repeat top-k collisions
         logits = torch.full_like(logits, float("-inf")).scatter_(-1, i, v)
+    # optionally crop the logits to smallest set of logits with a cumulative probability above top_p
+    if top_p is not None and top_p < 1.0:
```
Comment (carmocca) [inline litgpt/generate/base.py:37]: It would look nicer if it was refactored into a separate function. You can for example look at the Mistral implementation: https://github.com/mistralai/mistral-src/blob/8598cf582091a596671be31990448e0620017851/main.py#L12-L27

Comment (belerico) [reply litgpt/generate/base.py:37]: Sure. Do we also want to change the temperature scaling to be applied first as in HF? In the end one would have:

1. Temperature scaling
2. Top-k 
3. Top-p

Comment (carmocca) [reply litgpt/generate/base.py:37]: The order of (1) and (2) doesn't matter, right? It's only important that scaling is done before top-p. Having top-k first would also come with a tiny reduction in performance since you are trimming off the excess computations.

Comment (belerico) [reply litgpt/generate/base.py]: Yeah, you're right! I've updated the code with your suggestions

File: litgpt/chat/base.py [hunk 9]
```diff
@@ -33,7 +34,12 @@ def generate(
         prompt: Tensor of shape (T) with indices of the prompt sequence.
         max_returned_tokens: The maximum number of tokens to return (given plus generated).
         temperature: Scales the predicted logits by 1 / temperature
-        top_k: If specified, only sample among the tokens with the k highest probabilities
+        top_k: If specified, only sample among the tokens with the k highest probabilities.
+        top_p: The cumulative probability threshold to consider in the sampling process.
```
Comment (rasbt) [inline litgpt/chat/base.py:38]: It's clear to us, but I wonder if it would help newcomers to explicitly state that this should be a value between 0 and 1?

Comment (belerico) [reply litgpt/chat/base.py:38]: Something like this:

```python
"""
top_p: If specified, it represents the cumulative probability threshold to consider in the sampling process.
    In top-p sampling the next token is sampled from the highest probability tokens
    whose cumulative probability exceeds the threshold `top_p`. When specified
    it must be `0 <= top_p <= 1`, with `top_p=0` which is equivalent
    to sample the most probable token, while `top_p=1` samples from the whole distribution.
    It can be used in conjunction with `top_k` and `temperature` with the following order
    of application:

    1. `top_k`
    2. `temperature` scaling
    3. `top_p`

    For more details, see https://arxiv.org/abs/1904.09751
    or https://huyenchip.com/2024/01/16/sampling.html#top_p
"""
```

Comment (rasbt) [reply litgpt/chat/base.py:38]: Thanks, that sounds good to me. It's a bit long but I don't mind long if it's clear and helps users understand what's happening. I just have a small rewording but otherwise I like it! 

```
"""
top_p: If specified, it represents the cumulative probability threshold to consider in the sampling process.
    In top-p sampling, the next token is sampled from the highest probability tokens
    whose cumulative probability exceeds the threshold `top_p`. When specified,
    it must be `0 <= top_p <= 1`. Here, `top_p=0` is equivalent
    to sampling the most probable token, while `top_p=1` samples from the whole distribution.
    It can be used in conjunction with `top_k` and `temperature` with the following order
    of application:

    1. `top_k` sampling
    2. `temperature` scaling
    3. `top_p` sampling

    For more details, see https://arxiv.org/abs/1904.09751
    or https://huyenchip.com/2024/01/16/sampling.html#top_p
"""
```

File: litgpt/generate/base.py [hunk 10]
```diff
@@ -49,7 +49,9 @@ def sample(
         # do not use `torch.where` as in nanogpt because it will repeat top-k collisions
         logits = torch.full_like(logits, float("-inf")).scatter_(-1, i, v)
     # optionally scale the logits and sample from a probability distribution
-    if temperature > 0.0 and top_p is not None and top_p > 0.0:
+    if top_p is None:
```
Comment (rasbt) [inline litgpt/generate/base.py:52]: In that case, we could also just set it to 1.0 as the default. Any preference @carmocca ?

Comment (belerico) [reply litgpt/generate/base.py:52]: Yeah, I could set the default to `1.0` and check that is `0.0 <= top_p <= 1.0` and raise and error if it's not the case

Comment (carmocca) [reply litgpt/generate/base.py:52]: Agreed!

File: litgpt/generate/base.py [hunk 11]
```diff
@@ -24,16 +24,39 @@ def multinomial_num_samples_1(probs: torch.Tensor) -> torch.Tensor:
...
         v, i = torch.topk(logits, min(top_k, logits.size(-1)))
         # do not use `torch.where` as in nanogpt because it will repeat top-k collisions
         logits = torch.full_like(logits, float("-inf")).scatter_(-1, i, v)
     # optionally scale the logits and sample from a probability distribution
-    if temperature > 0.0:
-        probs = torch.nn.functional.softmax(logits / temperature, dim=-1)
+    if top_p is None:
+        top_p = 1.0
+    if temperature > 0.0 and top_p > 0.0:
+        logits = logits / temperature
```
Comment (carmocca) [inline litgpt/generate/base.py:55]: Shouldn't this be?:

```suggestion
    if temperature > 0.0 or top_p > 0.0:
        if temperature > 0.0:
            logits = logits / temperature
```

Comment (belerico) [reply litgpt/generate/base.py:55]: > Shouldn't this be?:

Yeah right i'm an idiot

File: litgpt/generate/base.py [hunk 12]
```diff
@@ -175,7 +231,7 @@ def main(
     L.seed_everything(1234)
     for i in range(num_samples):
         t0 = time.perf_counter()
-        y = generate(model, encoded, max_returned_tokens, temperature=temperature, top_k=top_k, eos_id=tokenizer.eos_id)
+        y = generate( model, encoded, max_returned_tokens, temperature=temperature, top_k=top_k, top_p=top_p, eos_id=tokenizer.eos_id)
```
Comment (rasbt) [inline litgpt/generate/base.py:234]:
```suggestion
        y = generate(model, encoded, max_returned_tokens, temperature=temperature, top_k=top_k, top_p=top_p, eos_id=tokenizer.eos_id)
```

File: tests/test_generate.py [hunk 13]
```diff
@@ -68,13 +68,13 @@ def test_main(fake_checkpoint_dir, monkeypatch, tensor_like):
...
     with redirect_stdout(out), redirect_stderr(err):
-        generate.main(temperature=2.0, top_k=2, num_samples=num_samples, checkpoint_dir=fake_checkpoint_dir)
+        generate.main(temperature=2.0, top_k=2, top_p=0.9, num_samples=num_samples, checkpoint_dir=fake_checkpoint_dir)
 
     assert len(tokenizer_mock.return_value.decode.mock_calls) == num_samples
     assert torch.allclose(tokenizer_mock.return_value.decode.call_args[0][0], generate_mock.return_value)
     assert (
         generate_mock.mock_calls
-        == [call(ANY, tensor_like, 53, temperature=2.0, top_k=2, eos_id=tokenizer_mock.return_value.eos_id)]
+        == [call(ANY, tensor_like, 53, temperature=2.0, top_k=2, top_p=0.9, eos_id=tokenizer_mock.return_value.eos_id)]
```
Comment (rasbt) [inline tests/test_generate.py:77]: Sorry, but I wonder if we should include one more test where `top_p` is changed so that it doesn't match the results anymore. Just as a sanity check that `top_p` has an effect.

**Comment by rasbt** [comment]
Thanks for all the updates and fixes. It looks all great to me now. The  only thing is perhaps adding one more unit test, but I can take care of that to make it easier.

**Review by rasbt — APPROVED** [review]

**Comment by belerico** [comment]
> Thanks for all the updates and fixes. It looks all great to me now. ~The only thing is perhaps adding one more unit test, but I can take care of that to make it easier.~ [done]

Thank you @rasbt: i had missed your comment

**Comment by rasbt** [comment]
No worries at all, I also thought it was probably quicker to just add instead of explain 😅

File: tests/test_generate.py [hunk 14]
```diff
@@ -102,10 +102,26 @@ def test_sample(temperature):
...
+    model.max_seq_length = 50
+    model.set_kv_cache(batch_size=1)
+
+    torch.manual_seed(123)
+    input_idx = torch.randint(10, size=(1,))
+
+    output1 = generate.generate(model, input_idx, 20, top_p=1.0)
+    output2 = generate.generate(model, input_idx, 20, top_p=0.1)
+
+    assert not torch.equal(output1, output2)
```
Comment (carmocca) [inline tests/test_generate.py:127]: This test is not useful because it will also pass if you set the same top_p. That's because `multinomial` advances the rng state.

For it to achieve the intended result, you need to seed before each call

```python
    torch.manual_seed(123)
    input_idx = torch.randint(10, size=(1,))

    torch.manual_seed(123)
    output1 = generate.generate(model, input_idx, 20, top_p=1.0)
    torch.manual_seed(123)
    output2 = generate.generate(model, input_idx, 20, top_p=0.1)
```

cc @rasbt

Comment (rasbt) [reply tests/test_generate.py:127]: arg, thanks!
